deeplearning4j-scaleout/spark/dl4j-spark-ml/src/main/scala/org/deeplearning4j/spark/sql/sources/mnist/MnistRelation.scala

```diff
@@ -30,36 +31,75 @@ case class MnistRelation(imagesPath: Path, labelsPath: Path)
   override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
     val sc = sqlContext.sparkContext
 
-    val manager = {
-      // CAVEAT: MnistManager supports only local files at this time.
-      val imagesFile = Paths.get(imagesPath.toUri).toFile.getAbsolutePath
-      val labelsFile = Paths.get(labelsPath.toUri).toFile.getAbsolutePath
-      new MnistManager(imagesFile, labelsFile)
-    }
-    try {
-      val cols = {
-        val req = requiredColumns.toSet
-        schema.fieldNames flatMap {
-          case "label" if req("label") => Seq(
-            (pt:Int) => manager.readLabel().toDouble)
-          case "features" if req("features") => Seq(
-            (pt:Int) => Vectors.dense(ArrayUtil.flatten(manager.readImage()).map(_.toDouble)))
-          case _ => Seq.empty
-        }
-      }
+    new MnistRDD(sc, imagesPath.toUri, labelsPath.toUri, requiredColumns, recordsPerPartition)
+  }
+}
+
+private class MnistPartition(override val index: Int, val startIndex: Int, val endIndex: Int)
+  extends Partition with Serializable {
+}
+
+private class MnistRDD(
+    sc: SparkContext,
+    private val imagesPath: URI,
+    private val labelsPath: URI,
+    private val requiredColumns: Array[String],
+    recordsPerPartition: Int = 1000
+  )
+  extends RDD[Row](sc, Nil) {
 
-      val numExamples = manager.getLabels.getCount
+  override val partitioner = None
 
-      sc.makeRDD((0 until numExamples).map(pt => {
```

Comment:
In future we must be wary of 'makeRDD', which pushes data from driver to executor.  Spark recommends that such a dataset be less than 100KB.